File: app/src/main/java/tpcreative/co/qrscanner/common/services/download/RetrofitInterface.java
package tpcreative.co.qrscanner.common.services.download;
import io.reactivex.Observable;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Response;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.POST;
import retrofit2.http.Streaming;
import retrofit2.http.Url;
import tpcreative.co.qrscanner.common.api.request.DownloadFileRequest;

/**
 * Created by PC on 9/1/2017.
 */

public interface RetrofitInterface {
    
    public static final String DOWNLOAD = "/api/file/download";


    @Streaming
    @GET
    Call<ResponseBody> downloadFileByUrl(@Url String fileUrl);

    // Retrofit 2 GET request for rxjava
    @Streaming
    @GET
    Observable<Response<ResponseBody>> downloadFileByUrlRx(@Url String fileUrl);
    
    @Streaming
    @POST(DOWNLOAD)
    Observable<Response<ResponseBody>> downloadFile(@Body DownloadFileRequest request);

    @Streaming
    @POST()
    Observable<Response<ResponseBody>> downloadFile(@Url String fileUrl, @Body DownloadFileRequest request);
    
    @GET()
    @Streaming
    Observable<Response<ResponseBody>> downloadFile(@Url String fileUrl);

    @GET(DOWNLOAD)
    @Streaming
    Observable<Response<ResponseBody>> downloadFile();

    @GET()
    @Streaming
    Observable<Response<ResponseBody>> downloadDriveFile(@Url String url, @Header("Authorization") String authToken);

}

File: app/src/main/java/tpcreative/co/qrscanner/common/services/download/RetrofitInterface.kt
package tpcreative.co.qrscanner.common.services.download

import io.reactivex.Observable
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Response
import retrofit2.http.*
import tpcreative.co.qrscanner.common.api.request.DownloadFileRequest

/**
 * Created by PC on 9/1/2017.
 */
interface RetrofitInterface {
    @Streaming
    @GET
    open fun downloadFileByUrl(@Url fileUrl: String?): Call<ResponseBody?>?

    // Retrofit 2 GET request for rxjava
    @Streaming
    @GET
    open fun downloadFileByUrlRx(@Url fileUrl: String?): Observable<Response<ResponseBody?>?>?
    @Streaming
    @POST(DOWNLOAD)
    open fun downloadFile(@Body request: DownloadFileRequest?): Observable<Response<ResponseBody?>?>?
    @Streaming
    @POST
    open fun downloadFile(@Url fileUrl: String?, @Body request: DownloadFileRequest?): Observable<Response<ResponseBody?>?>?
    @GET
    @Streaming
    open fun downloadFile(@Url fileUrl: String?): Observable<Response<ResponseBody?>?>?
    @GET(DOWNLOAD)
    @Streaming
    open fun downloadFile(): Observable<Response<ResponseBody?>?>?
    @GET
    @Streaming
    open fun downloadDriveFile(@Url url: String?, @Header("Authorization") authToken: String?): Observable<Response<ResponseBody?>?>?

    companion object {
        val DOWNLOAD: String? = "/api/file/download"
    }
}